examples.livemd

```diff
@@ -167,6 +167,140 @@ end)
+
+```elixir
+defmodule OpenAIWebSocket do
+  use WebSockex
+  require Logger
+
+  def start_link(opts) do
+    # OpenAI API docs: https://platform.openai.com/docs/guides/realtime
+    WebSockex.start_link(
+      "wss://api.openai.com/v1/realtime?model=gpt-4o-realtime-preview-2024-10-01",
+      __MODULE__,
+      %{response: <<>>},
+      extra_headers: [
+        {"Authorization", "Bearer " <> opts[:token]},
+        {"OpenAI-Beta", "realtime=v1"}
+      ]
+    )
+  end
+
+  def send_audio(ws, audio) do
+    audio = Base.encode64(audio)
+    frame = %{type: "input_audio_buffer.append", audio: audio} |> Jason.encode!()
+    WebSockex.send_frame(ws, {:text, frame})
+  end
+
+  def get_response_chunk(ws, chunk_byte_size) do
+    # There's no 'call' in WebSockex, so we just send and receive
+    send(ws, {:get_response_chunk, chunk_byte_size, self()})
+
+    receive do
+      {:response_chunk, chunk} -> chunk
+    end
+  end
+
+  @impl true
+  def handle_frame({:text, frame}, state) do
+    case Jason.decode!(frame) do
+      %{"type" => "response.audio.delta", "delta" => delta} ->
+        audio_payload = Base.decode64!(delta)
+        # Buffer the response audio
+        response = state.response <> audio_payload
+        {:ok, %{state | response: response}}
+
+      %{"type" => "input_audio_buffer.speech_started"} ->
+        # If the user speaks, they may interrupt the current response,
+        # so we drop it and wait for a new one.
+        {:ok, %{state | response: <<>>}}
+
+      %{"type" => "response.audio_transcript.done", "transcript" => transcript} ->
+        Logger.info("AI transcription: #{transcript}")
+        {:ok, state}
+
+      %{} = _event ->
+        {:ok, state}
+    end
+  end
+
+  @impl true
+  def handle_frame(_frame, state), do: {:ok, state}
+
+  @impl true
+  def handle_info({:get_response_chunk, size, pid}, state) do
+    case state.response do
+      <<chunk::binary-size(size), rest::binary>> ->
+        # If we have enough data, send it back
+        send(pid, {:response_chunk, chunk})
+        {:ok, %{state | response: rest}}
+
+      chunk ->
+        # Otherwise, send what we have, padded with silence
+        silence = <<0::size(size - byte_size(chunk))-unit(8)>>
+        send(pid, {:response_chunk, chunk <> silence})
+        {:ok, %{state | response: <<>>}}
+    end
+  end
+end
+```
+
+In the cell below, we receive stream from the browser via WebRTC, feed it to the API, receive response and send it back to the browser. You need to add the Open AI API token as a `OPEN_AI_TOKEN` secret in Livebook for this to wok. To connect via WebRTC, visit http://localhost:1234/talk_to_llm.html after running this cell
+
+```elixir
+{:ok, ws} = OpenAIWebSocket.start_link(token: System.fetch_env!("LB_OPEN_AI_TOKEN"))
```

Comment:
How about adding some check if `LB_OPEN_AI_TOKEN` is set and printing a warning in case it doesn't?

Reply:
Livebook does exactly this ;)